### PR TITLE
Add GeoDistanceCondition

### PR DESCRIPTION
**Pull request by ker0x**
This PR introduce `GeoDistanceCondition` and a new. `GeoPointField` for latitude and longitude coordinates:

### ToDos

- [x] Memory
- [x] Algolia
- [x] Elasticsearch
- [x] Loupe
   - [x] Edge Case: https://github.com/loupe-php/loupe/pull/80
   - [x] Edge Case: https://github.com/loupe-php/loupe/pull/82
- [x] Meilisearch
- [x] Opensearch
- [x] RedisSearch
   - [x] ~~https://github.com/RediSearch/RediSearch/issues/5020~~
- [x] Solr
- [x] Typesense

### CODE REVIEW

**Comment by alexander-schranz** [comment]
Thx for the pull request at current early state of the library we will not merge things into the core which are not part of all Adapters. Still we can keep this pull request open.

It would be great if you could create an issue for this also and if you have some time add links to the specific search engines documentation. I may would create 2 issues for GeoDistance and GeoBoundingBox as it are 2 seperate features/conditions and we may can only implement the once which are currently supported by all engines.

What we require is a list like this one:

https://github.com/schranz-search/schranz-search/issues/343#issuecomment-1917812158

**Comment by alexander-schranz** [comment]
@ker0x do you know if there is some GeoDistance Algorithm we could just copy paste into the [Memory adapter here](https://github.com/schranz-search/schranz-search/blob/d07492ee47c217699b7dcaa34ab5ff073acea1f0/packages/seal-memory-adapter/src/MemorySearcher.php#L151)?

For `Opensearch` I think the `Elasticsearch` implementation mostly can be copied?

File: docs/search-and-filters/index.rst [hunk 1]
```diff
@@ -186,6 +186,43 @@ the given value.
 
 The field is required to be marked as ``filterable`` in the index configuration.
 
+GeoDistanceCondition
+~~~~~~~~~~~~~~~~~~~~
+
+The ``GeoDistanceCondition`` is used to filter results within a radius by specifying a latitude, longitude and distance.
```
Comment (alexander-schranz) [inline docs/search-and-filters/index.rst]: What Unit does the distance use? we may should add it here and as a phpdoc to `GeoDistanceCondition`

File: packages/seal/src/Search/Condition/GeoDistanceCondition.php [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace Schranz\Search\SEAL\Search\Condition;
+
+class GeoDistanceCondition
+{
+    public function __construct(
+        public readonly string $field,
+        public readonly float $latitude,
+        public readonly float $longitude,
+        public readonly int|string $distance,
```
Comment (alexander-schranz) [inline packages/seal/src/Search/Condition/GeoDistanceCondition.php]: can we force a `int` here always?

**Comment by alexander-schranz** [comment]
I think we need also introduce a new `Type` for this .:

```php
return new Index('blog', [
    'id' => new Field\IdentifierField('id'),
    'geo' => new Field\GeoPointField('geo', filterable: true),
]);
```

This would be the first type without a map to a native PHP type so we need force how such information are converted into PHP raw format e.g.:

```php
'geo' => [
    'latitude' => 12.3,
    'longitude' => 43.8,
],
```

Basically we could use `lat` and `lon` but personally not a fan of using abbreviations.

We may need to have a look how different engine store it.

File: docs/search-and-filters/index.rst (hunk 1 above)
Comment (alexander-schranz) [reply docs/search-and-filters/index.rst]: added a test implementation based on `meters` int value.

File: packages/seal/src/Search/Condition/GeoDistanceCondition.php (hunk 2 above)
Comment (alexander-schranz) [reply packages/seal/src/Search/Condition/GeoDistanceCondition.php]: did currently go with `int` here

**Comment by alexander-schranz** [comment]
I added `GeoPointField` and added already a implementation for Memory Adapter and Meilisearch. Started with Loupe but did not yet found out why it doesn't show any results there 🤔

**Comment by ker0x** [comment]
Hello @alexander-schranz 

Sorry for the lack of feedback, I haven't had too much time available lately to continue working on this feature.

However, I've just added `GeoDistanceCondition` for Opensearch and fix `GeoDistanceCondition` for Solr and `GeoBoundingBoxCondition` for Opensearch.

**Comment by alexander-schranz** [comment]
No problem. Thank you for all your work you already did put into it. I pushed an update as I accidently added a regression, the Github CI is still a little bit flacky (checking the false commit out) but Meilisearch and Memory Adapter are now green with the GeoDistance test.

**Comment by alexander-schranz** [comment]
Thx to @toflar we solved the issues on Loupe side.
I pushed updates for Algolia and also Elasticsearch and Opensearch.
Typesense was also good to implement.

Missing for the GeoDistanceCondition is so only Solr and Redisearch now to update to the GeoPointField Type and make tests green.

**Comment by alexander-schranz** [comment]
Not sure if I did something wrong but there seems be an issue with the result returned by Redisearch: https://github.com/RediSearch/RediSearch/issues/5020.

**Comment by alexander-schranz** [comment]
Found the issue on my side lat <-> lon. So only Solr is last for the GeoDistanceCondition tests :)

**Comment by alexander-schranz** [comment]
As it seems the `GeoDistanceCondition` is now already finished I think we should move the `GeoBoundingBoxCondition` in its own pull request so we can merge this one:

<details>
<summary>Reapply GeoBoundingBoxCondition patch</summary>

```diff
diff --git a/docs/search-and-filters/index.rst b/docs/search-and-filters/index.rst
index b353b39..7861024 100644
--- a/docs/search-and-filters/index.rst
+++ b/docs/search-and-filters/index.rst
@@ -204,6 +204,22 @@ The ``GeoDistanceCondition`` is used to filter results within a radius by specif
 
 The field is required to be marked as ``filterable`` in the index configuration.
 
+GeoBoundingBoxCondition
+~~~~~~~~~~~~~~~~~~~~~~~
+
+The ``GeoBoundingBoxCondition`` is used to filter results within a bounding box by specifying a min latitude, min longitude, max latitude and max longitude.
+
+.. code-block:: php
+
+    <?php
+
+    use Schranz\Search\SEAL\Search\Condition;
+
+    $result = $this->engine->createSearchBuilder()
+        ->addIndex('restaurants')
+        ->addFilter(new Condition\GeoBoundingBoxCondition('location', 45.494181, 9.214024, 45.449484, 9.179175))
+        ->getResult();
+
 
 The field is required to be marked as ``filterable`` in the index configuration.
 
diff --git a/packages/seal-algolia-adapter/src/AlgoliaSearcher.php b/packages/seal-algolia-adapter/src/AlgoliaSearcher.php
index e5441bb..ee38380 100644
--- a/packages/seal-algolia-adapter/src/AlgoliaSearcher.php
+++ b/packages/seal-algolia-adapter/src/AlgoliaSearcher.php
@@ -106,6 +106,9 @@ final class AlgoliaSearcher implements SearcherInterface
                     ),
                     'aroundRadius' => $filter->distance,
                 ],
+                $filter instanceof Condition\GeoBoundingBoxCondition => $geoFilters = [
+                    'insideBoundingBox' => [[$filter->minLatitude, $filter->minLongitude, $filter->maxLatitude, $filter->maxLongitude]],
+                ],
                 default => throw new \LogicException($filter::class . ' filter not implemented.'),
             };
         }
diff --git a/packages/seal-meilisearch-adapter/src/MeilisearchSearcher.php b/packages/seal-meilisearch-adapter/src/MeilisearchSearcher.php
index 9b4657e..f99702f 100644
--- a/packages/seal-meilisearch-adapter/src/MeilisearchSearcher.php
+++ b/packages/seal-meilisearch-adapter/src/MeilisearchSearcher.php
@@ -92,6 +92,13 @@ final class MeilisearchSearcher implements SearcherInterface
                     $this->escapeFilterValue($filter->longitude),
                     $this->escapeFilterValue($filter->distance),
                 ),
+                $filter instanceof Condition\GeoBoundingBoxCondition => $filters[] = \sprintf(
+                    '_geoBoundingBox([%s, %s], [%s, %s])',
+                    $this->escapeFilterValue($filter->minLatitude),
+                    $this->escapeFilterValue($filter->minLongitude),
+                    $this->escapeFilterValue($filter->maxLatitude),
+                    $this->escapeFilterValue($filter->maxLongitude),
+                ),
                 default => throw new \LogicException($filter::class . ' filter not implemented.'),
             };
         }
diff --git a/packages/seal-opensearch-adapter/src/OpensearchSearcher.php b/packages/seal-opensearch-adapter/src/OpensearchSearcher.php
index 0acc16a..3380ad4 100644
--- a/packages/seal-opensearch-adapter/src/OpensearchSearcher.php
+++ b/packages/seal-opensearch-adapter/src/OpensearchSearcher.php
@@ -90,6 +90,16 @@ final class OpensearchSearcher implements SearcherInterface
                         'lon' => $filter->longitude,
                     ],
                 ],
+                $filter instanceof Condition\GeoBoundingBoxCondition => $query['bool']['filter']['geo_bounding_box'][$this->getFilterField($search->indexes, $filter->field)] = [
+                    'top_left' => [
+                        'lat' => $filter->minLatitude,
+                        'lon' => $filter->minLongitude,
+                    ],
+                    'bottom_right' => [
+                        'lat' => $filter->maxLatitude,
+                        'lon' => $filter->maxLongitude,
+                    ],
+                ],
                 default => throw new \LogicException($filter::class . ' filter not implemented.'),
             };
         }
diff --git a/packages/seal/src/Search/Condition/GeoBoundingBoxCondition.php b/packages/seal/src/Search/Condition/GeoBoundingBoxCondition.php
new file mode 100644
index 0000000..654a493
--- /dev/null
+++ b/packages/seal/src/Search/Condition/GeoBoundingBoxCondition.php
@@ -0,0 +1,26 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the Schranz Search package.
+ *
+ * (c) Alexander Schranz <alexander@sulu.io>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Schranz\Search\SEAL\Search\Condition;
+
+class GeoBoundingBoxCondition
+{
+    public function __construct(
+        public readonly string $field,
+        public readonly float $minLatitude,
+        public readonly float $minLongitude,
+        public readonly float $maxLatitude,
+        public readonly float $maxLongitude,
+    ) {
+    }
+}
```

</details>

**Comment by alexander-schranz** [comment]
@ker0x Merged. Thank you for your help here and kicking this off. If you want you can open the merge request for the `GeoBoundingBoxCondition` working branch.

**Pull request merged**